<!DOCTYPE html>
<h1>Wrapping errors</h1>
<br>
<p>An alternative to boxing errors is to wrap them in your own error type.</p>
<br>
<div data-lang="rust,editable"><div data-l="use std::error;"></div><div data-l="use std::num::ParseIntError;"></div><div data-l="use std::fmt;"></div><div data-l=""></div><div data-l="type Result&lt;T&gt; = std::result::Result&lt;T, DoubleError&gt;;"></div><div data-l=""></div><div data-l="#[derive(Debug)]"></div><div data-l="enum DoubleError {"></div><div data-l="    EmptyVec,"></div><div data-l="#    // We will defer to the parse error implementation for their error."></div><div data-l="#    // Supplying extra info requires adding more data to the type."></div><div data-l="    // ">We will defer to the parse error implementation for their error. Supplying extra info requires adding more data to the type.</div><div data-l="    Parse(ParseIntError),"></div><div data-l="}"></div><div data-l=""></div><div data-l="impl fmt::Display for DoubleError {"></div><div data-l="    fn fmt(&amp;self, f: &amp;mut fmt::Formatter) -&gt; fmt::Result {"></div><div data-l="        match *self {"></div><div data-l="            DoubleError::EmptyVec =&gt;"></div><div data-l="                write!(f, &quot;please use a vector with at least one element&quot;),"></div><div data-l="#            // This is a wrapper, so defer to the underlying types' implementation of `fmt`."></div><div data-l="            // ">This is a wrapper, so defer to the underlying types' implementation of <code>fmt</code>.</div><div data-l="            DoubleError::Parse(ref e) =&gt; e.fmt(f),"></div><div data-l="        }"></div><div data-l="    }"></div><div data-l="}"></div><div data-l=""></div><div data-l="impl error::Error for DoubleError {"></div><div data-l="    fn description(&amp;self) -&gt; &amp;str {"></div><div data-l="        match *self {"></div><div data-l="            DoubleError::EmptyVec =&gt; &quot;empty vectors not allowed&quot;,"></div><div data-l="#            // This already impls `Error`, so defer to its own implementation."></div><div data-l="            // ">This already impls <code>Error</code>, so defer to its own implementation.</div><div data-l="            DoubleError::Parse(ref e) =&gt; e.description(),"></div><div data-l="        }"></div><div data-l="    }"></div><div data-l=""></div><div data-l="    fn cause(&amp;self) -&gt; Option&lt;&amp;error::Error&gt; {"></div><div data-l="        match *self {"></div><div data-l="            DoubleError::EmptyVec =&gt; None,"></div><div data-l="#            // The cause is the underlying implementation error type. Is implicitly"></div><div data-l="#            // cast to the trait object `&amp;error::Error`. This works because the"></div><div data-l="#            // underlying type already implements the `Error` trait."></div><div data-l="            // ">The cause is the underlying implementation error type. Is implicitly cast to the trait object <code>&amp;error::Error</code>. This works because the underlying type already implements the <code>Error</code> trait.</div><div data-l="            DoubleError::Parse(ref e) =&gt; Some(e),"></div><div data-l="        }"></div><div data-l="    }"></div><div data-l="}"></div><div data-l=""></div><div data-l="#// Implement the conversion from `ParseIntError` to `DoubleError`."></div><div data-l="#// This will be automatically called by `?` if a `ParseIntError`"></div><div data-l="#// needs to be converted into a `DoubleError`."></div><div data-l="// ">Implement the conversion from <code>ParseIntError</code> to <code>DoubleError</code>. This will be automatically called by <code>?</code> if a <code>ParseIntError</code> needs to be converted into a <code>DoubleError</code>.</div><div data-l="impl From&lt;ParseIntError&gt; for DoubleError {"></div><div data-l="    fn from(err: ParseIntError) -&gt; DoubleError {"></div><div data-l="        DoubleError::Parse(err)"></div><div data-l="    }"></div><div data-l="}"></div><div data-l=""></div><div data-l="fn double_first(vec: Vec&lt;&amp;str&gt;) -&gt; Result&lt;i32&gt; {"></div><div data-l="    let first = vec.first().ok_or(DoubleError::EmptyVec)?;"></div><div data-l="    let parsed = first.parse::&lt;i32&gt;()?;"></div><div data-l=""></div><div data-l="    Ok(2 * parsed)"></div><div data-l="}"></div><div data-l=""></div><div data-l="fn print(result: Result&lt;i32&gt;) {"></div><div data-l="    match result {"></div><div data-l="        Ok(n)  =&gt; println!(&quot;The first doubled is {}&quot;, n),"></div><div data-l="        Err(e) =&gt; println!(&quot;Error: {}&quot;, e),"></div><div data-l="    }"></div><div data-l="}"></div><div data-l=""></div><div data-l="fn main() {"></div><div data-l="    let numbers = vec![&quot;42&quot;, &quot;93&quot;, &quot;18&quot;];"></div><div data-l="    let empty = vec![];"></div><div data-l="    let strings = vec![&quot;tofu&quot;, &quot;93&quot;, &quot;18&quot;];"></div><div data-l=""></div><div data-l="    print(double_first(numbers));"></div><div data-l="    print(double_first(empty));"></div><div data-l="    print(double_first(strings));"></div><div data-l="}"></div></div>
<br>
<p>This adds a bit more boilerplate for handling errors and might not be needed in
all applications. There are some libraries that can take care of the boilerplate
for you.</p>
<br>
<h3>See also:</h3>
<br>
<p><a class="notranslate" href="#3from"><code>From::from</code></a> and <a class="notranslate" href="#3enums"><code>Enums</code></a></p>
<br>
<a class="notranslate" href="#1https://doc.rust-lang.org/std/convert/trait.From.html">from</a>
<a class="notranslate" href="#1custom_types/enum.html">enums</a>